File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k%n;

        // reversing the whole array
        reverse(nums, 0, n-1);

        // reversing first k element
        reverse(nums, 0, k-1);

        // reversing the remaining part
        reverse(nums, k, n-1);
    }
    private void reverse(int[] arr, int si, int ei){
        while(si < ei){
            int temp = arr[si];
            arr[si] = arr[ei];
            arr[ei] = temp;
            si++;
            ei--;
        }
    }
}

File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k %= n;
        queue<int> q;
        for(int i=n-k;i<n;i++){
            q.push(nums[i]);
        }
        for(int i=0;i<n-k;i++){
            q.push(nums[i]);
        }
        for(int i=0;i<n;i++){
            nums[i] = q.front();
            q.pop();
        }
    }
};